ports/esp32/machine_i2s.c

```diff
@@ -0,0 +1,792 @@
+    { 6,  7, -1, -1, -1, -1, -1, -1 },  // Mono, 16-bits
+    { 4,  5,  6,  7, -1, -1, -1, -1 },  // Mono, 32-bits
+    { 6,  7,  2,  3, -1, -1, -1, -1 },  // Stereo, 16-bits
+    { 4,  5,  6,  7,  0,  1,  2,  3 },  // Stereo, 32-bits
+};
+
+//  The following function takes a sample buffer and swaps L/R channels
+//
+//  Background:  For 32-bit stereo, the ESP-IDF API has a L/R channel orientation that breaks
+//               convention with other ESP32 channel formats
+//
+//   appbuf[] = [L_0-7, L_8-15, L_16-23, L_24-31, R_0-7, R_8-15, R_16-23, R_24-31] = [Left channel, Right channel]
+//   dma[] =    [R_0-7, R_8-15, R_16-23, R_24-31, L_0-7, L_8-15, L_16-23, L_24-31] = [Right channel, Left channel]
+//
+//   where:
+//     L_0-7 is the least significant byte of the 32 bit sample in the Left channel
+//     L_24-31 is the most significant byte of the 32 bit sample in the Left channel
+//
+//  Example:
+//
+//   appbuf[] = [0x99, 0xBB, 0x11, 0x22, 0x44, 0x55, 0xAB, 0x77] = [Left channel, Right channel]
+//   dma[] =    [0x44, 0x55, 0xAB, 0x77, 0x99, 0xBB, 0x11, 0x22] = [Right channel,  Left channel]
+//   where:
+//      LEFT Channel =  0x99, 0xBB, 0x11, 0x22
+//      RIGHT Channel = 0x44, 0x55, 0xAB, 0x77
+//
+//    samples in appbuf are in little endian format:
+//       0x77 is the most significant byte of the 32-bit sample
+//       0x44 is the least significant byte of the 32-bit sample
+STATIC void swap_32_bit_stereo_channels(mp_buffer_info_t *bufinfo) {
+    int32_t swap_sample;
+    int32_t *sample = bufinfo->buf;
+    uint32_t num_samples = bufinfo->len / 4;
+    for (uint32_t i = 0; i < num_samples; i += 2) {
+        swap_sample = sample[i + 1];
+        sample[i + 1] = sample[i];
+        sample[i] = swap_sample;
+    }
+}
+
+STATIC int8_t get_frame_mapping_index(i2s_bits_per_sample_t bits, format_t format) {
+    if (format == MONO) {
+        if (bits == I2S_BITS_PER_SAMPLE_16BIT) {
+            return 0;
+        } else { // 32 bits
+            return 1;
+        }
+    } else { // STEREO
+        if (bits == I2S_BITS_PER_SAMPLE_16BIT) {
+            return 2;
+        } else { // 32 bits
+            return 3;
+        }
+    }
+}
+
+STATIC i2s_bits_per_sample_t get_dma_bits(uint8_t mode, i2s_bits_per_sample_t bits) {
+    if (mode == (I2S_MODE_MASTER | I2S_MODE_TX)) {
+        return bits;
+    } else { // Master Rx
+        // read 32 bit samples for I2S hardware.  e.g. MEMS microphones
+        return I2S_BITS_PER_SAMPLE_32BIT;
+    }
+}
+
+STATIC i2s_channel_fmt_t get_dma_format(uint8_t mode, format_t format) {
+    if (mode == (I2S_MODE_MASTER | I2S_MODE_TX)) {
+        if (format == MONO) {
+            return I2S_CHANNEL_FMT_ONLY_LEFT;
+        } else {  // STEREO
+            return I2S_CHANNEL_FMT_RIGHT_LEFT;
+        }
+    } else { // Master Rx
+        // read stereo frames for all I2S hardware
+        return I2S_CHANNEL_FMT_RIGHT_LEFT;
```

Comment:
This is great work.

Can you explain why you always read both left and right channels? It's quite unusual for people to have two microphones connected.

Reply:
You bring up a good point.  I need to take a second look at this implementation.   Valuable DMA buffer space is wasted for 16-bit and 32-bit Mono modes.  Thank you for reviewing this code and making suggestions.  I appreciate it !

Reply:
No problem - I'm really glad I found the PR and samples.